pkg/proxy/dns.go

```diff
@@ -0,0 +1,455 @@
+
+	handler.mu.Lock()
+	defer handler.mu.Unlock()
+
+	for i := range services {
+		service := &services[i]
+		name := types.NamespacedName{service.Namespace, service.Name}
+
+		if !api.IsServiceIPSet(service) {
+			glog.V(3).Infof("Skipping normal DNS entry for service %s due to clusterIP = %q", name, service.Spec.ClusterIP)
+			handler.isHeadless[name] = true
+			if entry, ok := handler.serviceRecords[name]; ok {
+				entry.confirm()
+			}
+			activeServices[name] = true
+			continue
+		}
+
+		handler.isHeadless[name] = false
+
+		handler.makeEntriesForService(name, service)
+		activeServices[name] = true
+	}
+
+	for name := range handler.serviceRecords {
+		if !activeServices[name] {
+			glog.V(3).Infof("Removing DNS entries for service %q", name)
+			delete(handler.serviceRecords, name)
+			delete(handler.isHeadless, name)
+		}
+	}
+}
+
+func (handler *DNSEndpointHandler) OnUpdate(endpointSets []api.Endpoints) {
+	handler.mu.Lock()
+	defer handler.mu.Unlock()
+
+	for i := range endpointSets {
+		endpointSet := &endpointSets[i]
+		name := types.NamespacedName{endpointSet.Namespace, endpointSet.Name}
+
+		isHeadless, confirmed := handler.isHeadless[name]
+		if confirmed && !isHeadless {
+			glog.V(3).Infof("Skipping non-headless service %s", name)
+			continue
+		}
+
+		handler.makeEntriesForEndpoints(name, endpointSet, confirmed)
+	}
+}
+
+func serviceSubdomain(serviceName types.NamespacedName, portName string, portProto api.Protocol, headlessId string) string {
+	if headlessId == "" {
+		if portName == "" && portProto == "" {
+			return strings.ToLower(fmt.Sprintf(srvPathFormat, serviceName.Name, serviceName.Namespace))
+		} else {
+			return strings.ToLower(fmt.Sprintf(srvPortPathFormat, portName, portProto, serviceName.Name, serviceName.Namespace))
+		}
+	} else {
+		if portName == "" && portProto == "" {
+			return strings.ToLower(fmt.Sprintf(headlessSrvPathFormat, headlessId, serviceName.Name, serviceName.Namespace))
+		} else {
+			return strings.ToLower(fmt.Sprintf(headlessSrvPortPathFormat, headlessId, portName, portProto, serviceName.Name, serviceName.Namespace))
+		}
+	}
+}
+
+func makeServicePortSpec(port api.ServicePort) string {
+	return fmt.Sprintf("_%s._%s", port.Name, strings.ToLower(string(port.Protocol)))
+}
+
+func makeEndpointPortSpec(port api.EndpointPort) string {
+	return fmt.Sprintf("_%s._%s", port.Name, strings.ToLower(string(port.Protocol)))
+}
+
+func (handler *DNSHandler) makeEntriesForService(name types.NamespacedName, service *api.Service) {
+	// the Group property lets us mimic the etcd behavior, since the "shortest" result will always
+	// be the "a" group, and thus only "a" records get returned unless we are asked for a SRV
+	// record by the full path
+	newEntry := normalServiceRecord{portRecords: make(map[string]skymsg.Service)}
+	srvPath := serviceSubdomain(name, "", "", "")
+	for _, port := range service.Spec.Ports {
+		newEntry.portRecords[makeServicePortSpec(port)] = skymsg.Service{
```

Comment:
You have several of these that set a bunch of default values. Use a constructor.